src/query_object.js

```diff
@@ -0,0 +1,228 @@
+'use strict';
+
+const util = require('util');
+const { isPlainObject } = require('./utils');
+const { parseExpr } = require('./expr');
+// deferred to break cyclic dependencies
+let Spell;
+
+const OPERATOR_MAP = {
+  $between: 'between',
+  $eq: '=',
+  $gt: '>',
+  $gte: '>=',
+  $in: 'in',
+  $like: 'like',
+  $lt: '<',
+  $lte: '<=',
+  $ne: '!=',
+  $nin: 'not in',
+  $notBetween: 'not between',
+  $notLike: 'not like',
+  $notIn: 'not in'
+};
+
+/**
+ * Parse object values as literal or subquery
+ * @param {Object} value
+ * @returns {Array<Object>}
+ */
+function parseValue(value) {
+  if (value instanceof Spell) return { type: 'subquery', value };
+  if (value && value.__raw) return { type: 'raw', value: value.value };
+  return parseExpr('?', value);
+}
+
+/**
+ * Check if object condition is an operator condition, such as `{ $gte: 100, $lt: 200 }`.
+ * @param {Object} condition
+ * @returns {boolean}
+ */
+function isOperatorCondition(condition) {
+  if (!isPlainObject(condition)) return false;
+  for (const $op in condition) {
+    if (OPERATOR_MAP.hasOwnProperty($op)) return true;
+  }
+  return false;
+}
+
+function merge(conditions, operator = 'and') {
+  return conditions.reduce((res, condition) => {
+    if (!res) return condition;
+    return { type: 'op', name: operator, args: [ res, condition ] };
+  }, null);
+}
+
+/**
+ * parse operator condition into expression ast
+ * @example
+ * parseOperator('id', { $gt: 0, $lt: 999999 });
+ * // => { type: 'op', name: 'and', args: [ ... ]}
+ * @param {string} name
+ * @param {Object} condition
+ * @returns {Object}
+ */
+function parseOperator(name, condition) {
+  const result = [];
+
+  for (const $op in condition) {
+    const operator = $op === '$not' ? OPERATOR_MAP.$ne : OPERATOR_MAP[$op];
```

Comment:
It suddenly occurred to me that how can we process `NOT` or `!` operator? eg.
```SQL
SELECT * FROM table_name
WHERE NOT NOT (id > 100);
```

Reply:
fixed with https://github.com/cyjake/leoric/pull/191/commits/664316be961ac8981a8c2774ad97b4b2f2fc65b1